File: [23.12]Design-Pattern/Java/src/structural/composite/NormalBox.java
package structural.composite;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class NormalBox implements Box {
    private final List<Item> list;
    private String name;
    private int price;

    public NormalBox(String name, int price) {
        this.name = name;
        this.price = price;
        this.list = new ArrayList<>();
    }

    @Override
    public void addItem(Item item) {
        list.add(item);
    }

    @Override
    public void removeItem(Item item) {
        list.remove(item);
    }

    @Override
    public int getAllPrice() {
        return list.stream()
                .mapToInt(item -> item instanceof Box box ? box.getAllPrice() + item.getPrice() : item.getPrice())
                .sum();
    }

    @Override
    public int getPrice() {
        return this.price;
    }

    @Override
    public String getName() {
        return this.name;
    }

    @Override
    public String getItems() {
        return getName() + " = { " + list.stream().map(item -> item instanceof Box box ? box.getItems() : item.getName()).collect(Collectors.joining(", ")) + " }";
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/structural/composite/NormalBox.kt
package structural.composite

class NormalBox(private val name: String, private val price: Int) : Box {
    private val list = mutableListOf<Item>()

    override fun getAllPrice(): Int = list.sumOf {
        when (it) {
            is Box -> it.getAllPrice() + it.getPrice()
            else -> it.getPrice()
        }
    }

    override fun getItems(): String = "$name = { ${list.joinToString(", ") { item ->
        when (item) {
            is Box -> item.getItems()
            else -> item.getName()
        }
    }} }"

    override fun addItem(item: Item) {
        list.add(item)
    }

    override fun removeItem(item: Item) {
        list.remove(item)
    }

    override fun getPrice(): Int = price

    override fun getName(): String = name
}
